Added in version specification for app download tracking

seancunningham opened the pull request:
Sorry if there was a better way to update the pull request. Changes have been made as per your requests in #95. The version is now passed in via the constructor, not a new method as before, and a unit test has been made.

d4rken commented:
> Sorry if there was a better way to update the pull request.
Normally you would squash or revert your commits, then force push your changes towards your pullrequest branch. Google has more info on that.

I think setting the version via constructor is slightly restricting here. What about a method  `setVersion(@Nullable String version)` and then in the respective places of `DownloadTracker` make use of a `getVersion()` that pulls the version from the `mVersion` field if it is non null, and otherwise pulls it from the apps context. This would nice integration within the TrackHelper.

Make an extra test method for the above case.

I think then we can merge it, I might a small code polish PR, and then the current dev branch including your changes can be merged with master and we'll do a new release.

d4rken commented on piwik-sdk/src/main/java/org/piwik/sdk/DownloadTracker.java:
```diff
@@ -50,20 +57,23 @@ public DownloadTracker(Tracker tracker, TrackMe baseTrackMe) {
+            mVersion = Integer.toString(pkgInfo.versionCode);   
+            mName = pkgInfo.packageName;
+        }
+        catch (PackageManager.NameNotFoundException e) {
```
No new line here.

d4rken commented on piwik-sdk/src/test/java/org/piwik/sdk/DownloadTrackerTest.java:
```diff
@@ -86,7 +86,22 @@ public void testTrackNewAppDownload() throws Exception {
+        downloadTracker.trackNewAppDownload(DownloadTracker.Extra.APK_CHECKSUM);
+        Thread.sleep(100); // APK checksum happens off thread
+        queryParams = parseEventUrl(tracker.getLastEvent());
+        checkNewAppDownload(queryParams);
```
As extra test method please.

dotsbb commented on piwik-sdk/src/main/java/org/piwik/sdk/DownloadTracker.java:
```diff
@@ -41,7 +43,25 @@
+    
+    public void setVersion(String version) {
+        mVersion = version;
+        System.out.println("Version is now "+version);
```
I would suggest to use `Logy.d` here instead, as we did it [here](https://github.com/piwik/piwik-sdk-android/blob/master/piwik-sdk/src/main/java/org/piwik/sdk/Tracker.java#L723).

dotsbb commented on piwik-sdk/src/main/java/org/piwik/sdk/DownloadTracker.java:
```diff
@@ -50,20 +70,25 @@ public DownloadTracker(Tracker tracker, TrackMe baseTrackMe) {
+            PackageInfo pkgInfo = mPackMan.getPackageInfo(mContext.getPackageName(), 0);
+            mVersion = getVersion();
+            if (mVersion == null){
+                throw new PackageManager.NameNotFoundException("Version could not be found");
```
This `try/catch` block and those one above at `L56` looks very similar. It's better to refactor them and follow DRY principle

seancunningham commented:
Totally right, the try/catches were a bit unnecessary other than to get the package (which I don't think is necessary anymore, but followed the pattern I saw). I've refactored it to only have the try/catch in the constructor - as long as the version getter is using the same pkgInfo object, the try/catch isn't needed. Also removed that debug line -- wasn't intending to commit that and am glad it was caught.

d4rken commented on piwik-sdk/src/main/java/org/piwik/sdk/DownloadTracker.java:
```diff
@@ -50,20 +68,26 @@ public DownloadTracker(Tracker tracker, TrackMe baseTrackMe) {
         mPackMan = mContext.getPackageManager();
+        try {
+            mPkgInfo = mPackMan.getPackageInfo(mPackageName, 0);
+            mName = mPkgInfo.packageName;
```
The packagename is already stored in `mPackageName`.

d4rken commented on piwik-sdk/src/main/java/org/piwik/sdk/DownloadTracker.java:
```diff
@@ -50,20 +68,26 @@ public DownloadTracker(Tracker tracker, TrackMe baseTrackMe) {
+                throw new PackageManager.NameNotFoundException("Version could not be found");
+            }
+        }
+        catch (PackageManager.NameNotFoundException e) {
```
No new line.

d4rken commented on piwik-sdk/src/main/java/org/piwik/sdk/DownloadTracker.java:
```diff
@@ -41,7 +44,22 @@
+        if (mVersion != null){
+            return mVersion;
+        }
+        else if (mPkgInfo != null){
```
No new line.

d4rken commented on piwik-sdk/src/main/java/org/piwik/sdk/DownloadTracker.java:
```diff
@@ -41,7 +44,22 @@
+    }
+    
+    public String getVersion(){
+        if (mVersion != null){
```
Missing space between the brackets.

d4rken commented on piwik-sdk/src/main/java/org/piwik/sdk/DownloadTracker.java:
```diff
@@ -50,20 +68,26 @@ public DownloadTracker(Tracker tracker, TrackMe baseTrackMe) {
-                    mPreferences.edit().putBoolean(firedKey, true).apply();
-                    trackNewAppDownload(extra);
-                }
+        String firedKey = "downloaded:" + mName + ":" + mVersion;
```
The idea is to use getVersion() through out the code instead of mVersion, which load either provide a manually set version or pull one from the context.

d4rken commented on piwik-sdk/src/main/java/org/piwik/sdk/DownloadTracker.java:
```diff
@@ -50,20 +68,26 @@ public DownloadTracker(Tracker tracker, TrackMe baseTrackMe) {
+            mPkgInfo = mPackMan.getPackageInfo(mPackageName, 0);
+            mName = mPkgInfo.packageName;
+            mVersion = getVersion();
+            if (mVersion == null){
```
If `getPackageInfo` did not throw an exception then `getVersion()` is never null. A packageInfo object always has a `versionCode`. The null check and extra exception is superfluos.

dotsbb commented on piwik-sdk/src/test/java/org/piwik/sdk/DownloadTrackerTest.java:
```diff
@@ -99,6 +100,39 @@ public void testTrackNewAppDownload() throws Exception {
+        Thread.sleep(100); // APK checksum happens off thread
+        queryParams = parseEventUrl(tracker.getLastEvent());
+        checkNewAppDownload(queryParams);
+       m = REGEX_DOWNLOADTRACK.matcher(queryParams.get(QueryParams.DOWNLOAD));
```
Indentation to fix

dotsbb commented:
:+1: Looks good to me
